### PR TITLE
[Training] Add distributed checkpointing

### PR DESCRIPTION
**Pull request by kevin314**
TODO:

- [ ]  ~~Separate method to get safetensors weights from the saved distributed checkpoint for larger models?~~
- [x]  Setting seed makes validation videos consistent, may also be affecting training (investigate further)

Saves training checkpoint with weights, optimizer states, stateful dataloader, and scheduler for resuming training as distcp files.
Also saves a consolidated checkpoint as a safetensors file that can be used for inference.

Mostly reuses finetrainer's checkpointing

### CODE REVIEW

**Comment by Edenzzzz** [comment]
I think we actually want to try Bytecheckpoint for this?

File: fastvideo/v1/training/training_utils.py [hunk 1]
```diff
@@ -16,6 +22,26 @@
...
+    for param_name, param in sharded_sd.items():
+        if param.is_cpu:
+            # Move back to device if offloaded to CPU
+            param = param.to(device)
+        if hasattr(param, "_local_tensor"):
+            # Gather DTensor
+            param = param.full_tensor()
+        if rank <= 0:
+            cpu_state_dict[param_name] = param.cpu()
+        torch.distributed.barrier()
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: `full_tensor` calls [all-gather which is blocking](https://github.com/pytorch/pytorch/blob/c014e4bcaa1f507864dfb650d7136bb2ab2d3d77/torch/distributed/tensor/placement_types.py#L251). Repeatedly calling barrier is unnecessary.

File: fastvideo/v1/training/training_utils.py [hunk 2]
```diff
@@ -66,24 +92,67 @@ def get_sigmas(noise_scheduler,
...
+                rank,
+                end_time - begin_time,
+                local_main_process_only=False)
+
+    cpu_state = gather_state_dict_on_cpu_rank0(transformer,
+                                               device=None,
+                                               is_main_process=True)
 
-    # Save it (only on rank 0 since we used rank0_only=True)
     if rank <= 0:
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: Use `rank == 0`

File: fastvideo/v1/training/training_utils.py [hunk 3]
```diff
@@ -66,24 +92,67 @@ def get_sigmas(noise_scheduler,
...
 
-    # Now get the state dict
-    cpu_state = transformer.state_dict()
+    if dataloader is not None:
+        states["dataloader"] = dataloader
+
+    if scheduler is not None:
+        states["scheduler"] = SchedulerWrapper(scheduler)
+
+    distcp_dir = os.path.join(save_dir, "distributed_checkpoint")
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: for alignment with offcial style, we may want to rename all `distcp` to `dcp`

File: fastvideo/v1/training/training_utils.py [hunk 4]
```diff
@@ -94,6 +163,64 @@ def save_checkpoint(transformer, rank, output_dir, step) -> None:
...
+    Returns the step number from which training should resume.
+    """
+    if not os.path.exists(checkpoint_path):
+        logger.warning("Checkpoint path %s does not exist", checkpoint_path)
+        return 0
+
+    # Extract step number from checkpoint path
+    step = int(os.path.basename(checkpoint_path).split('-')[-1])
+
+    if rank <= 0:
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: `rank == 0`

File: fastvideo/v1/training/training_utils.py [hunk 5]
```diff
@@ -16,6 +22,26 @@
...
+    cpu_state_dict = {}
+    sharded_sd = model.state_dict()
+    for param_name, param in sharded_sd.items():
+        if param.is_cpu:
+            # Move back to device if offloaded to CPU
+            param = param.to(device)
+        if hasattr(param, "_local_tensor"):
+            # Gather DTensor
+            param = param.full_tensor()
+        if rank <= 0:
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: `rank == 0`

File: fastvideo/v1/training/training_utils.py [hunk 6]
```diff
@@ -16,6 +22,26 @@
...
+                                   device: Optional[torch.device] = None,
+                                   *,
+                                   is_main_process: bool) -> Dict[str, Any]:
+    rank = int(os.environ.get("RANK", 0))
+    cpu_state_dict = {}
+    sharded_sd = model.state_dict()
+    for param_name, param in sharded_sd.items():
+        if param.is_cpu:
+            # Move back to device if offloaded to CPU
+            param = param.to(device)
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: If it's on cpu, can we just gather on cpu? I think gathering on CPU uses IPC + DRAM, which is faster than CPU <-> GPU via PCIE

File: fastvideo/v1/training/training_utils.py [hunk 7]
```diff
@@ -16,6 +22,26 @@
 _HAS_ERRORED_CLIP_GRAD_NORM_WHILE_HANDLING_FAILING_DTENSOR_CASES = False
 
 
+def gather_state_dict_on_cpu_rank0(model,
+                                   device: Optional[torch.device] = None,
+                                   *,
+                                   is_main_process: bool) -> Dict[str, Any]:
+    rank = int(os.environ.get("RANK", 0))
```
Comment (Edenzzzz) [inline fastvideo/v1/training/training_utils.py]: `dist.get_rank()` is more reliable

**Review by Edenzzzz — CHANGES_REQUESTED** [review]

**Comment by Edenzzzz** [comment]
Better get rid of the [rank=-1 logic here](https://github.com/hao-ai-lab/FastVideo/blob/7768bb80f618e670617cd0d440d91ab4fd558333/fastvideo/v1/distributed/parallel_state.py#L773) in the future

**Comment by kevin314** [comment]
> I think we actually want to try Bytecheckpoint for this?

We looked into the library and it doesn't seem to be maintained that actively, so we wanted to try finetrainer's implementation first

**Comment by kevin314** [comment]
Ready for re-review

**Review by Edenzzzz — APPROVED** [review]
LGTM

**Review by SolitaryThinker — APPROVED** [review]
